services/s3/src/main/java/software/amazon/awssdk/services/s3/parsing/S3Uri.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.parsing;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * Object that represents a parsed S3 URI. Can be used to easily retrieve the bucket, key, region, style, and query parameters
+ * of the URI. Only path-style and virtual-hosted-style URI parsing is supported, including CLI-style URIs, e.g.,
+ * "s3://bucket/key". AccessPoints and Outposts URI parsing is not supported. If you work with object keys and/or query
+ * parameters with special characters, they must be URL-encoded, e.g., replace " " with "%20". If you work with
+ * virtual-hosted-style URIs with bucket names that contain a dot, i.e., ".", the dot must not be URL-encoded. Encoded buckets,
+ * keys, and query parameters will be returned decoded.
+ */
+@Immutable
+@SdkPublicApi
+public final class S3Uri implements ToCopyableBuilder<S3Uri.Builder, S3Uri> {
+
+    private final URI uri;
+    private final String bucket;
+    private final String key;
+    private final Region region;
+    private final boolean isPathStyle;
+    private final Map<String, List<String>> queryParams;
+
+    private S3Uri(Builder builder) {
+        this.uri = Validate.notNull(builder.uri, "URI must not be null");
+        this.bucket = builder.bucket;
+        this.key = builder.key;
+        this.region = builder.region;
+        this.isPathStyle = Validate.notNull(builder.isPathStyle, "Path style flag must not be null");
+        this.queryParams = Validate.notNull(builder.queryParams, "Query parameters map must not be null");
```

Comment:
It may be a better experience if we don't require null here and make it default empty map.
this.queryParams = queryParams == null ? new HashMap<>() : queryParams;

Reply:
+1